tests/quantum_coroutine_state_handler_tests.cpp

```diff
@@ -0,0 +1,228 @@
+#include <quantum_fixture.h>
+
+using namespace Bloomberg;
+using namespace quantum;
+using ms = std::chrono::milliseconds;
+
+namespace {
+
+struct TaskParams
+{
+    size_t yieldIterations{0}; // number of times the task is supposed to yield
+    bool randomYieldIterations{false}; // if true the task will yield a random number of times in range [0, yieldIterations]
+    ms sleepTime{30}; // sleep time between yield calls. If yieldIterations is 0, a task will sleep sleepTime ms
+    bool randomSleepTime{false}; // if true the task will sleep random time in range [0, sleepTime]
+    bool throwException{false}; // indicates if a task should throw exception
+    size_t exceptionIteration{0}; // iteration on which exception is thrown
+    ITask::RetCode returnCode{ITask::RetCode::Success}; // return code of the task
+};
+
+std::function<int(CoroContext<int>::Ptr)> makeTask(const TaskParams& taskParms)
+{
+    auto yield = [](CoroContext<int>::Ptr ctx) -> bool {
+        ctx->yield();
+        return true;
+    };
+    return [taskParms, yield](CoroContext<int>::Ptr ctx)-> int {
+        size_t iteration = 0;
+        size_t yieldIterations = taskParms.yieldIterations;
+        if (taskParms.randomYieldIterations)
+        {
+            yieldIterations = rand() % (taskParms.yieldIterations + 1);
+        }
+
+        do {
+            ms sleepTime = taskParms.sleepTime;
+            if (taskParms.randomSleepTime)
+            {
+                sleepTime = static_cast<ms>(rand() % taskParms.sleepTime.count() + 1);
+            }
+            std::this_thread::sleep_for(sleepTime);
+            if (taskParms.throwException && (taskParms.exceptionIteration == iteration))
+            {
+                throw std::runtime_error("Unexpected error");
+            }
+        } while ((iteration++ < yieldIterations) && yield(ctx));
+
+        return static_cast<int>(taskParms.returnCode);
+    };
+}
+
+class CoroutineStateHandlerTest : public ::testing::TestWithParam<CoroutineStateHandler>
+{
+protected:
+    CoroutineStateHandlerTest()
+    {
+        srand((unsigned) time(NULL));
+    }
+
+    void testCoroutineStateHandler(
+        size_t tasksCount,
+        const TaskParams& taskParams,
+        bool coroutineSharingForAny = false,
+        bool loadBalanceSharedIoQueues = false)
+    {
+        CoroutineStateHandler finalStateHandler{};
+        auto& coroutineStateHandler = GetParam();
+        if (coroutineStateHandler)
+        {
+            finalStateHandler = [this, coroutineStateHandler](CoroutineState state)
+            {
+                countCoroutineStates(state);
+                coroutineStateHandler(state);
+            };
+        }
+        const TestConfiguration config(loadBalanceSharedIoQueues,
+                                       coroutineSharingForAny,
+                                       finalStateHandler);
+        auto dispatcher = DispatcherSingleton::createInstance(config);
+        dispatcher->drain();
+        {
```

Comment:
No need in this local scope anymore. I previously used it to estimate execution time